### ISSUE TITLE
PS/Chief Nodes not terminating & Worker Nodes not accurately terminating with tf.estimator.train_and_evaluate

### ISSUE THREAD

**Issue by wuciawe** [tensorflow/tensorflow#21305]
I'm using tf.estimator.train_and_evaluate of TF 1.9.

When the training reaches the `max_steps` and `max_steps` is less than the dataset's size, the evaluator and the non-chief workers will terminate as expected. But, the ps nodes and the chief worker do not terminate automatically. They just stuck, and you have to kill them manually.

When the `max_steps` is larger than the dataset's size, only one of the workers will terminate, which I guess is the one that reaches the end of the dataset. The remaining workers and ps nodes stuck.

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce
Mobile device

**Comment by tensorflowbutler** [comment]
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by wuciawe** [comment]
It is.

**Comment by rohan100jain** [comment]
Can you describe your setup a bit more? How many workers / ps nodes, any sample code / minimal example you can provide so that we can reproduce and debug it?

**Comment by wuciawe** [comment]
@rohan100jain I make a minimal example for linear regression. First, run `python gen_data.py` to generate data. Then, run 

`python model.py --ps_hosts=127.0.0.1:2222 --chief_hosts=127.0.0.1:2223 --worker_hosts=127.0.0.1:2224 --task_index=0 --job_name=ps`

`python model.py --ps_hosts=127.0.0.1:2222 --chief_hosts=127.0.0.1:2223 --worker_hosts=127.0.0.1:2224 --task_index=0 --job_name=chief`

`python model.py --ps_hosts=127.0.0.1:2222 --chief_hosts=127.0.0.1:2223 --worker_hosts=127.0.0.1:2224 --task_index=0 --job_name=worker`

`python model.py --ps_hosts=127.0.0.1:2222 --chief_hosts=127.0.0.1:2223 --worker_hosts=127.0.0.1:2224 --task_index=0 --job_name=evaluator`

the code is as follows:

gen_data.py
```
import random
import tensorflow as tf
import os

if __name__ == "__main__":
    if not os.path.exists(os.path.join('.', 'data')):
        os.makedirs(os.path.join('.', 'data'))
    train_out = tf.python_io.TFRecordWriter('data/train.tfrecord')
    for _ in range(10000):
        x1 = (random.randint(1, 10000) - 5000) / 10000.0
        x2 = (random.randint(1, 10000) - 5000) / 10000.0
        x3 = (random.randint(1, 10000) - 5000) / 10000.0
        y = 3 * x1 - 2 * x2 + 5 * x3
        example = tf.train.Example(features=tf.train.Features(
            feature={
                'x': tf.train.Feature(float_list=tf.train.FloatList(value=[x1, x2, x3])),
                'y': tf.train.Feature(float_list=tf.train.FloatList(value=[y])),
            }
        ))
        train_out.write(example.SerializeToString())
    train_out.close()
    test_out = tf.python_io.TFRecordWriter('data/test.tfrecord')
    for _ in range(1000):
        x1 = (random.randint(1, 10000) - 5000) / 10000.0
        x2 = (random.randint(1, 10000) - 5000) / 10000.0
        x3 = (random.randint(1, 10000) - 5000) / 10000.0
        y = 3 * x1 - 2 * x2 + 5 * x3
        example = tf.train.Example(features=tf.train.Features(
            feature={
                'x': tf.train.Feature(float_list=tf.train.FloatList(value=[x1, x2, x3])),
                'y': tf.train.Feature(float_list=tf.train.FloatList(value=[y])),
            }
        ))
        test_out.write(example.SerializeToString())
    test_out.close()

```

model.py
```
import tensorflow as tf
import os
import json

FLAGS = tf.app.flags.FLAGS
tf.app.flags.DEFINE_string("ps_hosts", '', "Comma-separated list of hostname:port pairs")
tf.app.flags.DEFINE_string("chief_hosts", '', "Comma-separated list of hostname:port pairs")
tf.app.flags.DEFINE_string("worker_hosts", '', "Comma-separated list of hostname:port pairs")
tf.app.flags.DEFINE_string("job_name", '', "One of 'ps', 'worker', 'chief', 'evaluator'")
tf.app.flags.DEFINE_integer("task_index", 0, "Index of task within the job")


def input_fn(file_dir, num_epochs, num_workers=1, task_index=0):
    def _parse_fn(record):
        features = {
            "x": tf.FixedLenFeature([3], tf.float32),
            "y": tf.FixedLenFeature([1], tf.float32)
        }
        parsed = tf.parse_single_example(record, features)
        label = parsed.pop("y")
        return parsed, label
    dataset = tf.data.Dataset.list_files(file_dir, shuffle=False)
    dataset = dataset.apply(tf.contrib.data.parallel_interleave(tf.data.TFRecordDataset, cycle_length=1))
    dataset = dataset.map(_parse_fn, num_parallel_calls=32).prefetch(10)
    dataset = dataset.apply(tf.contrib.data.shuffle_and_repeat(buffer_size=4, count=num_epochs))
    if num_workers > 1:
        dataset = dataset.shard(num_workers, task_index)
    dataset = dataset.batch(4)
    iterator = dataset.make_one_shot_iterator()
    batch_features, batch_labels = iterator.get_next()
    return batch_features, batch_labels


def set_dist_env():
    ps_hosts = FLAGS.ps_hosts.split(',')
    chief_hosts = FLAGS.chief_hosts.split(',')
    worker_hosts = FLAGS.worker_hosts.split(',')
    task_index = FLAGS.task_index
    job_name = FLAGS.job_name
    tf_config = {
        'cluster': {'chief': chief_hosts, 'worker': worker_hosts, 'ps': ps_hosts},
        'task': {'type': job_name, 'index': task_index}
    }
    print(json.dumps(tf_config))
    os.environ['TF_CONFIG'] = json.dumps(tf_config)


def model_fn(features, labels, mode, params=None):
    x = features["x"]
    w = tf.get_variable(name='w', dtype=tf.float32, shape=[3, 1], initializer=tf.glorot_normal_initializer())
    b = tf.get_variable(name='b', dtype=tf.float32, shape=[1], initializer=tf.constant_initializer(0.0))
    pred = tf.add(b, tf.matmul(x, w))
    predictions = {"prediction": pred}
    if mode == tf.estimator.ModeKeys.PREDICT:
        export_outputs = {
            tf.saved_model.signature_constants.DEFAULT_SERVING_SIGNATURE_DEF_KEY: tf.estimator.export.PredictOutput(predictions)
        }
        return tf.estimator.EstimatorSpec(mode=mode, predictions=predictions, export_outputs=export_outputs)
    error = tf.reduce_mean(tf.square(labels - pred))
    if mode == tf.estimator.ModeKeys.EVAL:
        return tf.estimator.EstimatorSpec(mode=mode, predictions=predictions, loss=error)
    if mode == tf.estimator.ModeKeys.TRAIN:
        train_op = tf.train.GradientDescentOptimizer(0.001).minimize(error, global_step=tf.train.get_global_step())
        return tf.estimator.EstimatorSpec(mode=mode, predictions=predictions, loss=error, train_op=train_op)


def main(_):
    tr_dir = 'data/train*'
    va_dir = 'data/test*'
    set_dist_env()
    config = tf.estimator.RunConfig().replace(
        session_config=tf.ConfigProto(device_count={'GPU': 0, 'CPU': 32}),
        log_step_count_steps=10000,
        save_summary_steps=10000,
        save_checkpoints_steps=10000)
    e = tf.estimator.Estimator(model_fn=model_fn, model_dir="model", params={}, config=config)
    num_workers = len(FLAGS.worker_hosts.split(",")) + 1
    worker_index = 0
    if FLAGS.job_name == "worker":
        worker_index = int(FLAGS.task_index) + 1
    train_spec = tf.estimator.TrainSpec(input_fn=lambda: input_fn(tr_dir, num_epochs=9999, num_workers=num_workers, task_index=worker_index), max_steps=200000)
    feature_spec = {
        "x": tf.FixedLenFeature([3], tf.float32),
    }
    serving_input_receiver_fn = tf.estimator.export.build_parsing_serving_input_receiver_fn(feature_spec)
    exporters = tf.estimator.LatestExporter("linear", serving_input_receiver_fn)
    eval_spec = tf.estimator.EvalSpec(
        input_fn=lambda: input_fn(va_dir, num_epochs=1),
        steps=1000,
        start_delay_secs=5,
        throttle_secs=10,
        exporters=exporters)
    tf.estimator.train_and_evaluate(e, train_spec, eval_spec)


if __name__ == "__main__":
    tf.logging.set_verbosity(tf.logging.INFO)
    tf.app.run()

```

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by weberxie** [comment]
@wuciawe  I didn't reproduce this problem with your code in TF r1.10.

**Comment by wuciawe** [comment]
@weberxie Do you mean the ps, the chief, and the worker all will terminate properly when the training step reaches the max_steps?

**Comment by weberxie** [comment]
@wuciawe Only the ps still running when step reaches the max_steps.

**Comment by wuciawe** [comment]
@weberxie So the problem remains.

**Comment by weberxie** [comment]
@wuciawe PS does not terminate is normal.

**Comment by jinyuhang** [comment]
@ Have you killled the ps when using the estimator for the distributed tensorflow? I met the same problem,
I try many ways, but it was still failed to shutdowm the ps.

I view the code of the estimator. I only find the ps server is "join()".  but i want to close the PS

thank you

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 31 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 46 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by zhjunqin** [comment]
We met same issue in TF 1.12 like above described.

set max_step in TrainSepc to 126800

```python
    model = build_estimator(build_model_column)

    train_spec = tf.estimator.TrainSpec(input_fn=train_input_fn, max_steps=FLAGS.max_step)
    eval_spec = tf.estimator.EvalSpec(input_fn=eval_input_fn, steps=6000, throttle_secs=120)
    for index in range(Config.TRAIN_EPOCHS):
        tf.estimator.train_and_evaluate(model, train_spec, eval_spec)
        tf.logging.info('=' * 60)
```
Then after 126800 steps, the whole training stop, but `chief` still not exit
```
INFO:tensorflow:Saving checkpoints for 1268001 into hdfs://xxx
```

We test same code in TF 1.7 and TF 1.8, chief can exit as expect, but in TF 1.10 and TF 1.12, chief didn't exit even max_step reached.

**Comment by zhjunqin** [comment]
We just tested TF 1.13 and found that chief can exit after max_step reached.

**Comment by byronyi** [comment]
> @weberxie Interesting. How would you think the lack of automatic resource cleanup is normal for a machine learning framework in production. In a model training workflow, how to notify the workflow controller the termination of the training stage to terminate the remaining PSs and to update the trained model to online service in a programmable way?

It is expected that the cluster resource manager would terminate PS. It could be a programmable framework like Apache Yarn or Kubernetes, or some scripts that kills PS after worker exits, or even a personnel who just does the job manually.

**Comment by wuciawe** [comment]
> > @weberxie Interesting. How would you think the lack of automatic resource cleanup is normal for a machine learning framework in production. In a model training workflow, how to notify the workflow controller the termination of the training stage to terminate the remaining PSs and to update the trained model to online service in a programmable way?
> 
> It is expected that the cluster resource manager would terminate PS. It could be a programmable framework like Apache Yarn or Kubernetes, or some scripts that kills PS after worker exits, or even a personnel who just does the job manually.

I thought distributed tf was/would itself be that kind of framework, similar to that spark/flink can recover from failure and terminate after job finished in standalone mode without the help of YARN/K8S.

PS Do the job manually is not expected for production usage.

**Comment by byronyi** [comment]
I’m sorry, but apparently TF is not designed around the way you expected.

**Comment by wangdejian** [comment]
We met same issue in TF 1.10 like above described, but no problem in TF 1.13

**Comment by Ningshiqi** [comment]
I met the same issue in TF 1.14. 
I have 2 ps ,2 worker .Woker1 is finished successfully. But "worker0" is chief node , stuck in `Saving checkpoints for 10002 into hdfs:*****model.ckpt.`.Ps0 is the evaluate node, is finished successfully,ps1 is stuck in ` 2020-10-16 09:49:03.411776: I tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc:365] Started server with target: grpc://localhost:2222
20/10/16 01:50:32 WARN util.NativeCodeLoader main: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable`

**Comment by Ningshiqi** [comment]
> I met the same issue in TF 1.14.
> I have 2 ps ,2 worker .Woker1 is finished successfully. But "worker0" is chief node , stuck in `Saving checkpoints for 10002 into hdfs:*****model.ckpt.`.Ps0 is the evaluate node, is finished successfully,ps1 is stuck in ` 2020-10-16 09:49:03.411776: I tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc:365] Started server with target: grpc://localhost:2222 20/10/16 01:50:32 WARN util.NativeCodeLoader main: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable`

I try the same code in TF.1.12, it works!

**Comment by sushreebarsa** [comment]
Hi There,

We are moving this issue to closed status, as you are using an older version of tensorflow(1.x) which is officially considered as end of life. We recommend that you upgrade to 2.4 or later version and let us know if the issue still persists in newer versions.
Please open a new issue for any help you need against 2.x, and we will get you the right help.Thanks!
